### PR TITLE
GODRIVER-1490 Split parsing and validation logic

### PR DESCRIPTION
**Pull request by tfogo**
Removes validation logic from ConnString.parse(). Validation
is still run as part of connstring.Parse(), but there is now a
connstring.ParseWithoutValidating() function that can be used if
parsing without validation is required. ConnString.Validate() can
be used to run validation separately.

### CODE REVIEW

**Comment by tfogo** [comment]
There are still some test failures, but they don't look related to this PR.

**Comment by tfogo** [comment]
I replaced all references to `Parse()` with `ParseAndValidate()` to ensure there are no changes in behavior.

File: x/mongo/driver/connstring/connstring.go [hunk 1]
```diff
@@ -21,7 +21,24 @@ import (
...
+// It check that all values are valid.
+func ParseAndValidate(s string) (ConnString, error) {
+	p := parser{dnsResolver: dns.DefaultResolver}
+	err := p.parse(s)
+	if err != nil {
+		return p.ConnString, internal.WrapErrorf(err, "error parsing uri")
+	}
+	err = p.ConnString.Validate()
+	if err != nil {
+		return p.ConnString, internal.WrapErrorf(err, "error parsing uri")
```
Comment (divjotarora) [inline x/mongo/driver/connstring/connstring.go]: Should this error message be "error validating uri" instead of "error parsing uri"?

File: x/mongo/driver/connstring/connstring.go [hunk 2]
```diff
@@ -21,7 +21,24 @@ import (
 	"go.mongodb.org/mongo-driver/x/mongo/driver/wiremessage"
 )
 
-// Parse parses the provided uri and returns a URI object.
+// ParseAndValidate parses the provided uri and returns a URI object.
```
Comment (divjotarora) [inline x/mongo/driver/connstring/connstring.go]: If we're making changes, I'd prefer for this to say "ParseAndValidate parses the provided URI into a ConnString object" because the terminology and capitalization in the current docs is not correct. Similar changes required to `Parse` below.

File: x/mongo/driver/connstring/connstring.go [hunk 3]
```diff
@@ -21,7 +21,24 @@ import (
...
+	}
+	err = p.ConnString.Validate()
+	if err != nil {
+		return p.ConnString, internal.WrapErrorf(err, "error parsing uri")
+	}
+	return p.ConnString, nil
+}
+
+// Parse parses the provided uri and returns a URI object
+// but does not check that all values are valid. Use `ConnString.ParseAndValidate()`
```
Comment (divjotarora) [inline x/mongo/driver/connstring/connstring.go]: Was this supposed to say "Use `ConnString.Validate` to run validation checks separately"?

File: x/mongo/driver/connstring/connstring.go [hunk 4]
```diff
@@ -37,7 +54,9 @@ type ConnString struct {
 	AppName                            string
 	AuthMechanism                      string
 	AuthMechanismProperties            map[string]string
+	AuthMechanismPropertiesSet         bool
```
Comment (divjotarora) [inline x/mongo/driver/connstring/connstring.go]: Are these changes needed? You can check if a map is set via `map != nil`.

File: x/mongo/driver/topology/sdam_spec_test.go [hunk 5]
```diff
@@ -110,7 +110,8 @@ func (r *response) UnmarshalJSON(buf []byte) error {
 func setUpFSM(t *testing.T, uri string) *fsm {
 	fsm := newFSM()
 
-	cs, err := connstring.Parse(uri)
+
```
Comment (divjotarora) [inline x/mongo/driver/topology/sdam_spec_test.go]: nit: Can you remove the extra newline that was added?

File: x/mongo/driver/connstring/connstring.go (hunk 4 above)
Comment (tfogo) [reply x/mongo/driver/connstring/connstring.go]: Yes because AuthMechanismProperties can be set to a default map in `parser.setDefaultAuthParams()`. I'm using `AuthMechanismPropertiesSet` to check if the `AuthMechanismProperties` had been set before we apply a default.

Comment (divjotarora) [reply x/mongo/driver/connstring/connstring.go]: Makes sense. Resolving.

File: x/mongo/driver/connstring/connstring.go (hunk 1 above)
Comment (tfogo) [reply x/mongo/driver/connstring/connstring.go]: Yeah this was a shortcut to keep all the tests passing that expect an error that says `"error parsing uri"`. If you think it's important that we differentiate the errors, I can go through the tests and update them.

File: x/mongo/driver/connstring/connstring.go (hunk 3 above)
Comment (tfogo) [reply x/mongo/driver/connstring/connstring.go]: yep

File: x/mongo/driver/connstring/connstring.go (hunk 1 above)
Comment (divjotarora) [reply x/mongo/driver/connstring/connstring.go]: This is likely fine as we had this error for both pieces before too.

File: mongo/integration/initial_dns_seedlist_discovery_test.go [hunk 6]
```diff
@@ -62,13 +62,13 @@ func runSeedlistDiscoveryTest(mt *mtest.T, file string) {
...
 
-	cs, err := connstring.Parse(test.URI)
+	cs, err := connstring.ParseAndValidate(test.URI)
 	if test.Error {
 		assert.NotNil(mt, err, "expected URI parsing error, got nil")
 		return
 	}
 	// the resolved connstring may not have valid credentials
-	if err != nil && err.Error() == "error parsing uri: authsource without username is invalid" {
+	if err != nil && err.Error() == "error validating uri: authsource without username is invalid" {
```
Comment (iwysiu) [inline mongo/integration/initial_dns_seedlist_discovery_test.go]: You can change line 65 to just use connstring.Parse and remove this if statement

Comment (tfogo) [reply mongo/integration/initial_dns_seedlist_discovery_test.go]: Oh nice! Just pushed that change.
